rueidisotel/trace.go

```diff
@@ -17,93 +17,122 @@ import (
 var (
 	name   = "github.com/redis/rueidis"
 	kind   = trace.WithSpanKind(trace.SpanKindClient)
-	tracer = otel.Tracer(name)
-	meter  = otel.Meter(name)
 	dbattr = attribute.String("db.system", "redis")
-
-	cscMiss, _ = meter.Int64Counter("rueidis_do_cache_miss")
-	cscHits, _ = meter.Int64Counter("rueidis_do_cache_hits")
 )
 
 var _ rueidis.Client = (*otelclient)(nil)
 
-// WithClient creates a new rueidis.Client with OpenTelemetry tracing enabled
+// WithClient creates a new rueidis.Client with OpenTelemetry tracing enabled.
 func WithClient(client rueidis.Client, opts ...Option) rueidis.Client {
-	o := &otelclient{client: client}
-	for _, fn := range opts {
-		fn(o)
+	o := &otelclient{
+		client: client,
+	}
+	for _, opt := range opts {
+		opt(o)
+	}
+	if o.meterProvider == nil {
+		o.meterProvider = otel.GetMeterProvider() // Default to global MeterProvider
+	}
+	if o.tracerProvider == nil {
+		o.tracerProvider = otel.GetTracerProvider() // Default to global TracerProvider
 	}
+	// Now that we have the meterProvider and tracerProvider, get the Meter and Tracer
+	o.meter = o.meterProvider.Meter(name)
+	o.tracer = o.tracerProvider.Tracer(name)
+	// Now create the counters using the meter
+	o.cscMiss, _ = o.meter.Int64Counter("rueidis_do_cache_miss")
+	o.cscHits, _ = o.meter.Int64Counter("rueidis_do_cache_hits")
 	return o
 }
 
 // Option is the Functional Options interface
 type Option func(o *otelclient)
 
-// MetricAttrs set additional attributes to append to each metric.
-func MetricAttrs(attrs ...attribute.KeyValue) Option {
+// WithMetricAttrs MetricAttrs set additional attributes to append to each metric.
+func WithMetricAttrs(attrs ...attribute.KeyValue) Option {
 	return func(o *otelclient) {
 		o.mAttrs = attrs
 	}
 }
 
-// TraceAttrs set additional attributes to append to each trace.
-func TraceAttrs(attrs ...attribute.KeyValue) Option {
+// WithTraceAttrs set additional attributes to append to each trace.
+func WithTraceAttrs(attrs ...attribute.KeyValue) Option {
```

Comment:
Oh, this is a breaking change. We can’t do that directly. We have at least to maintain an alias to the original function name.

Reply:
Got it